File: ch03/src/bin/sequentially_consistent_ordering.rs
use std::{
    sync::atomic::{AtomicBool, Ordering},
    thread,
    time::Duration,
};

static A: AtomicBool = AtomicBool::new(false);
static B: AtomicBool = AtomicBool::new(false);

static mut S: String = String::new();

fn main() {
    const DELAY: bool = false;
    let a = thread::spawn(|| {
        // Warn the other thread that this thread is about to access S
        A.store(true, Ordering::SeqCst);

        if DELAY {
            thread::sleep(Duration::from_millis(20));
        }

        // Check that the other thread hasn't access S before changing S
        if !B.load(Ordering::SeqCst) {
            println!("Thread a: B has not been set to true, -> accessing S");
            thread::sleep(Duration::from_millis(20));
            unsafe {
                S.push('!');
            }
        }
    });

    let b = thread::spawn(|| {
        // Warn the other thread that this thread is about to access S
        B.store(true, Ordering::SeqCst);

        if DELAY {
            thread::sleep(Duration::from_millis(20));
        }

        // Check that the other thread hasn't access S before changing S
        if !A.load(Ordering::SeqCst) {
            println!("Thread b: A has not been set to true, -> accessing S");
            unsafe { S.push('!') };
        }
    });
    a.join().unwrap();
    b.join().unwrap();

    // How does it work?
    // The SeqCst ordering ensure a global single order.
    // In any scenario, the first operation is a store operation, which prevents the other
    // thread from accessing S.
    // Thus there is at most 1 thread can access S.
    // (There are scenarios in which the first 2 operations are stores, in which no thread accesses S.)

    unsafe {
        if S.is_empty() {
            println!("No thread accessed S, Ok");
        } else if S.len() == 1 {
            println!("One thread access S, Ok");
        } else {
            eprintln!("More than 1 thread accessed S -> FAIL!");
            std::process::exit(1);
        }
    }
}

File: ch03/src/cpp/sequentially_consistent_ordering.cpp
#include <atomic>
#include <chrono>
#include <cstdlib>
#include <iostream>
#include <string>
#include <thread>

namespace {
static auto sA = std::atomic_bool { false };
static auto sB = std::atomic_bool { false };
static auto sS = std::atomic_int32_t { 0 };
}

int main()
{
    static const auto kDelay { false };

    auto t1 = std::thread { [] {
        sA.store(true, std::memory_order_seq_cst);
        if (kDelay) {
            std::this_thread::sleep_for(std::chrono::milliseconds { 20 });
        }

        if (!sB.load(std::memory_order_seq_cst)) {
            std::cout << "Thread a: B has not been set to true, we will now access S\n";
            std::this_thread::sleep_for(std::chrono::milliseconds { 20 });
            sS += 1;
        }
    } };

    auto t2 = std::thread { [] {
        sB.store(true, std::memory_order_seq_cst);
        if (kDelay) {
            std::this_thread::sleep_for(std::chrono::milliseconds { 20 });
        }

        if (!sA.load(std::memory_order_seq_cst)) {
            std::cout << "Thread b: A has not been set to true, we will now access S\n";
            std::this_thread::sleep_for(std::chrono::milliseconds { 20 });
            sS += 1;
        }
    } };

    t1.join();
    t2.join();

    if (sS == 0) {
        std::cout << "No thread accessed S, Ok\n";
    } else if (sS == 1) {
        std::cout << "One thread accessed S, Ok\n";
    } else {
        std::cout << "More than 1 thread accessed S -> FAIL!\n";
        std::abort();
    }

    return 0;
}
